commands/http/client.go

```diff
@@ -264,23 +228,91 @@ func getResponse(httpRes *http.Response, req cmds.Request) (cmds.Response, error
 
 		res.SetError(e, e.Code)
 
-	} else {
-		outputType := reflect.TypeOf(req.Command().Type)
-		var v interface{}
-
-		if outputType != nil {
-			v = reflect.New(outputType).Interface()
-			err = dec.Decode(v)
-		} else {
-			err = dec.Decode(&v)
+		return res, nil
+	}
+
+	outputType := reflect.TypeOf(req.Command().Type)
+	v, err := decodeTypedVal(outputType, dec)
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+
+	res.SetOutput(v)
+
+	return res, nil
+}
+
+// read json objects off of the given stream, and write the objects out to
+// the 'out' channel
+func readStreamedJson(req cmds.Request, httpRes *http.Response, out chan<- interface{}) {
+	defer close(out)
+	dec := json.NewDecoder(&httpResponseReader{httpRes})
+	outputType := reflect.TypeOf(req.Command().Type)
+
+	ctx := req.Context()
+
+	for {
+		v, err := decodeTypedVal(outputType, dec)
+		if err != nil {
+			// since we are just looping reading on the response, the only way to
+			// know we are 'done' is for the consumer to close the response body.
+			// doing so doesnt throw an io.EOF, but we want to treat it like one.
+			if !(strings.Contains(err.Error(), "read on closed response body") || err == io.EOF) {
+				log.Error(err)
+			}
+			return
 		}
-		if err != nil && err != io.EOF {
-			return nil, err
+
+		select {
+		case <-ctx.Done():
+			return
+		case out <- v:
 		}
-		if v != nil {
-			res.SetOutput(v)
+
+	}
+}
```

Comment:
okay, a little thought and i have an idea:

instead of having this `<-chan interface{}` thing everywhere, we should have an object that wraps that channel, and an error channel, and simply has a method `GetNextObject(ctx) (interface{}, error)` on it that selects on the context, the output and the error. It should clean up a lot of the logic